core/src/sumeragi/main_loop.rs

```diff
@@ -635,479 +725,276 @@ pub fn run<F>(
 
-        let current_view_change_index: u64 = {
+        let current_view_change_index = {
             prune_view_change_proofs_and_calculate_current_index(
                 &state,
                 &mut view_change_proof_chain,
             )
         };
 
-        compare_view_change_index_and_block_height_to_old(
-            sumeragi,
+        reset_state(
+            addr,
             current_view_change_index,
             &mut old_view_change_index,
             state.latest_block_height,
             &mut old_latest_block_height,
             &mut state.current_topology,
-            &mut voting_block_option,
-            &mut block_signature_acc,
-            &mut has_sent_transactions,
-            &mut instant_when_we_should_create_a_block,
-        );
-
-        handle_role_agnostic_messages(
-            sumeragi,
-            &mut state,
-            &mut maybe_incoming_message,
-            &mut voting_block_option,
+            &mut voting_block,
+            &mut voting_signatures,
+            &mut round_start_time,
+            &mut view_change_suggested,
         );
 
-        if state.current_topology.role(&sumeragi.peer_id) != Role::Leader {
-            if !state.transaction_cache.is_empty() && !has_sent_transactions {
-                // It is assumed that we only need to send 1 tx to check liveness.
-                let tx = state
-                    .transaction_cache
-                    .choose(&mut rand::thread_rng())
-                    .expect("It was checked earlier that transaction cache is not empty.")
-                    .clone();
-                let tx_hash = tx.hash();
-                info!(
-                    peer_addr = %sumeragi.peer_id.address,
-                    peer_role = ?state.current_topology.role(&sumeragi.peer_id),
-                    leader_addr = %state.current_topology.leader().address,
-                    %tx_hash,
-                    "Forwarding tx to leader"
-                );
-
-                // Don't require leader to submit receipts and therefore create blocks if the tx is still waiting for more signatures.
-                if matches!(tx.check_signature_condition(&state.wsv), Ok(MustUse(true))) {
-                    sumeragi.broadcast_packet_to(
-                        MessagePacket::new(
-                            view_change_proof_chain.clone(),
-                            crate::sumeragi::TransactionForwarded::new(tx).into(),
-                        ),
-                        std::iter::once(state.current_topology.leader()),
-                    );
-                    has_sent_transactions = true;
-                    sent_transaction_time = Instant::now();
-                }
-            }
+        if round_start_time.elapsed() > sumeragi.pipeline_time() && !view_change_suggested {
+            // NOTE: Either there are malicious peers or there are no transactions.
+            // There is no meaningful performance hit when there are no transaction.
```

Comment:
The peer should not suggest a view change if it has no local transactions it wants to see get through. And the view change suggestions are not guaranteed to be delivered. In order to ensure a view change does eventually happen the suspecting peers must periodically call suggest_view_change() until the view change occurs.

Reply:
> The peer should not suggest a view change if it has no local transactions it wants to see get through

why not, what's the harm? Other peers will request it anyhow. I'm trying to keep it as simple as possible

> And the view change suggestions are not guaranteed to be delivered. In order to ensure a view change does eventually happen the suspecting peers must periodically call suggest_view_change() until the view change occurs.

got you, I believe it's fixed now

Reply:
ok, this is a valid point. Let me see what I can do about it

Reply:
I've addressed it with comments in the code. If you don't mind I prefer the view change is made at least every `pipeline_time` if there is no block committed(for whatever reason) or earlier if the block is committed

Reply:
We can do it that way. We can always change back later if in production it is confusing.

Reply:
I made it so that it doesn't look confusing in the logs. Otherwise, I would have to special case and tangle up the code